### PR TITLE
[Executor] Support dataclass and primitive types in aggregation

### PR DESCRIPTION
**Pull request by lumoslnt**
# Description

Currently, the input of flex flow aggregation function is the output dictionary in line result, not the original output of `__call__` function, so in this PR, we support passing the data class and primitive types output directly for aggregation function.

## Main Changes:

1. [`src/promptflow-core/promptflow/executor/_line_execution_process_pool.py`](https://github.com/microsoft/promptflow/pull/2947/files#diff-d49c96161ef5c6f24cd63defe125d750cf1b78391f29c5fe3be165ca1c1ac778): Previously, we convert both data class and primitive types to dictionary, which make it difficult for batch engine to determine whether the original output is a dictionary or a primitive type, so in this PR, we only convert data class to dictionary here.
2. [`src/promptflow-devkit/promptflow/batch/_batch_engine.py`](diffhunk://#diff-0c95dd2aaf088a46d489e437605bbaa71150175508828fc5bbd36b5e15f4b554L488-R497): Based on the changes in `_line_execution_process_pool.py`, the aggregation function can get the original output of primitive types, but for data class, it is still a dictionary, so we convert the dictionary to `AttrDict` here, which support the attribute access to its key. Although, it is still not a real data class, user can access the dictionary in attribute access way.

Other changes:

* [`src/promptflow-core/promptflow/_utils/dataclass_serializer.py`](diffhunk://#diff-c084667ef87ee52a4d091449a9ca63f7e09ff6b91d224b31839b0b072ca06395R94-R100): Introduced a new function `convert_dataclass_to_dict` that converts data classes to dictionaries. This function is used in place of `convert_eager_flow_output_to_dict` in other parts of the codebase.
* [`src/promptflow-core/promptflow/contracts/types.py`](diffhunk://#diff-61ff2296efe62c3cff11d5d160fa9bafa4c0554e453c723411362c91dcc2b53aL6-R6): Added a new `AttrDict` class that allows attribute access to dictionary keys. This class is used to wrap dictionary outputs in other parts of the codebase.

Test files:

* `src/promptflow-devkit/tests/sdk_cli_test/e2etests/test_flow_test.py`, `src/promptflow-core/tests/core/e2etests/test_eager_flow.py`: Modified the test assertions to use attribute-style access instead of dictionary-style access. [[1]](diffhunk://#diff-162af09d07468e1340c2e7487f5964c9c7a3a54ac6430f11429a788d09e597c4L58-R64) [[2]](diffhunk://#diff-b125543ffc0d08b4dd6bf8d327df9bc6cfc3c58d3696fe031871d2346e11eb94L450-R454) [[3]](diffhunk://#diff-b125543ffc0d08b4dd6bf8d327df9bc6cfc3c58d3696fe031871d2346e11eb94L471-R479)
* `src/promptflow/tests/test_configs/eager_flows/basic_callable_class/simple_callable_class.py`, `src/promptflow/tests/test_configs/eager_flows/basic_callable_class_async/simple_callable_class.py`, `src/promptflow/tests/test_configs/eager_flows/basic_callable_class_with_sample_file/simple_callable_class.py`: Replaced `TypedDict` with `dataclass` for the `FlowOutput` class and modified the `__call__` and `__aggregate__` methods to use attribute-style access. [[1]](diffhunk://#diff-4716dff409b5c0d9e16289d4fc2a71d9c10d2db9d57df56afc7c9540cb67fe8fL4-R10) [[2]](diffhunk://#diff-4716dff409b5c0d9e16289d4fc2a71d9c10d2db9d57df56afc7c9540cb67fe8fL20-R29) [[3]](diffhunk://#diff-09a0d40f2482c844262875e28c42f5cd66aad65756fac7bcca68164d4695a340R4-R10) [[4]](diffhunk://#diff-09a0d40f2482c844262875e28c42f5cd66aad65756fac7bcca68164d4695a340L19-R30) [[5]](diffhunk://#diff-de7d0c3daa427b4d0e32b7cc90f9c0138ae4c467e0973d01855d49e602f96504L4-R11) [[6]](diffhunk://#diff-de7d0c3daa427b4d0e32b7cc90f9c0138ae4c467e0973d01855d49e602f96504L20-R30)

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/tests/test_configs/eager_flows/basic_callable_class/simple_callable_class.py [hunk 1]
```diff
@@ -17,13 +19,14 @@ def __init__(self, obj_input: str):
...
-        }
+        return FlowOutput(obj_input=self.obj_input, func_input=func_input, obj_id=id(self))
 
     def __aggregate__(self, results: list) -> dict:
+        # Try attribute-style access for the datacalss
+        obj_inputs = [r.obj_input for r in results]
+        func_inputs = [r.func_input for r in results]
+        obj_ids = [r.obj_id for r in results]
+
         return {"length": len(results)}
```
Comment (thy09) [inline src/promptflow/tests/test_configs/eager_flows/basic_callable_class/simple_callable_class.py:30]: Do we cover the scenario that return result is a raw string?

Comment (lumoslnt) [reply src/promptflow/tests/test_configs/eager_flows/basic_callable_class/simple_callable_class.py:30]: Added.

File: src/promptflow-core/promptflow/contracts/types.py [hunk 2]
```diff
@@ -54,3 +54,16 @@ def serialize(self):
...
+
+
+class AttrDict(dict):
+    """A dictionary that allows attribute access to its keys."""
+
+    def __getattr__(self, key: str) -> Any:
+        if key in self:
+            return self[key]
+        else:
+            raise AttributeError(f"'{self.__class__.__name__}' object has no attribute '{key}'")
```
Comment (thy09) [inline src/promptflow-core/promptflow/contracts/types.py:66]: you can simply use super().__getattr__

**Comment by thy09** [comment]
Add more description about the overall logic of this change.

File: src/promptflow-core/promptflow/contracts/types.py (hunk 2 above)
Comment (lumoslnt) [reply src/promptflow-core/promptflow/contracts/types.py]: Simply using `return self[key]` here.

**Review by thy09 — APPROVED** [review]

File: src/promptflow/tests/test_configs/eager_flows/basic_callable_class_with_sample_file/simple_callable_class.py [hunk 3]
```diff
@@ -1,11 +1,14 @@
...
 # ---------------------------------------------------------
-from typing import TypedDict
+
+from dataclasses import dataclass
 
 from promptflow.tracing import trace
 
-class FlowOutput(TypedDict):
+
+@dataclass
```
Comment (D-W-) [inline src/promptflow/tests/test_configs/eager_flows/basic_callable_class_with_sample_file/simple_callable_class.py:10]: Consider add a new test case for dataclass output? We need to support both TypedDict and dataclass.

**Review by D-W- — APPROVED** [review]

File: src/promptflow/tests/test_configs/eager_flows/basic_callable_class_with_sample_file/simple_callable_class.py (hunk 3 above)
Comment (lumoslnt) [reply src/promptflow/tests/test_configs/eager_flows/basic_callable_class_with_sample_file/simple_callable_class.py:10]: Will fix in other PR.

**Pull request merged**